trino-aws-proxy-spark4/src/main/java/io/trino/aws/proxy/spark4/PresignAwareS3Client.java

```diff
@@ -0,0 +1,118 @@
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+import software.amazon.awssdk.services.s3.model.S3Request;
+import software.amazon.awssdk.services.s3.model.UploadPartRequest;
+import software.amazon.awssdk.services.s3.model.UploadPartResponse;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class PresignAwareS3Client
+        extends DelegatingS3Client
+{
+    private static final Log log = LogFactory.getLog(PresignAwareS3Client.class);
+
+    public PresignAwareS3Client(S3Client delegate)
+    {
+        super(delegate);
+    }
+
+    private Optional<URI> getPresignedUrl(String method, String bucket, String key, Optional<UploadPartRequest> uploadPartRequest)
+    {
+        Optional<String> uploadId = uploadPartRequest.map(UploadPartRequest::uploadId);
+        Optional<Integer> partNumber = uploadPartRequest.map(UploadPartRequest::partNumber);
+
+        log.debug("request.get-presigned-url method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber));
+
+        Map<String, List<String>> headers;
+        try {
+            HeadObjectResponse response = headObject(builder -> {
+                builder.bucket(bucket).key(key);
+                partNumber.ifPresent(builder::partNumber);
+                uploadId.ifPresent(id -> builder.overrideConfiguration(c -> c.putRawQueryParameter("uploadId", id)));
+            });
+            headers = response.sdkHttpResponse().headers();
+        }
+        catch (AwsServiceException e) {
+            boolean is404NotFound = (e.statusCode() == 404);
+            boolean isGetOrDelete = method.equalsIgnoreCase("get") || method.equalsIgnoreCase("delete");
+
+            if (is404NotFound && isGetOrDelete) {
+                throw e;
+            }
+
+            // when the bucket isn't found an exception is thrown - but response headers still have pre-signed URLs
+            // given that this not GET/DELETE, we can return the pre-signed URL
+            headers = e.awsErrorDetails().sdkHttpResponse().headers();
+        }
+        catch (RuntimeException e) {
+            log.error("request.get-presigned-url.exception method=%s bucket=%s key=%s uploadId=%s partNumber=%s".formatted(method, bucket, key, uploadId, partNumber), e);
+            throw e;
+        }
+
+        List<String> headerValue = headers.get("X-Trino-Pre-Signed-Url-" + method);
+        log.debug("response.get-presigned-url.exception method=%s bucket=%s key=%s uploadId=%s partNumber=%s success=%s".formatted(method, bucket, key, uploadId, partNumber, (headerValue != null) && !headerValue.isEmpty()));
+        return ((headerValue != null) && !headerValue.isEmpty()) ? Optional.of(URI.create(String.valueOf(headerValue.get(0)))) : Optional.empty();
+    }
+
+    private <T extends S3Request.Builder> T registerPresignedUrl(T builder, URI presignedUrl)
```

Comment:
why not make this get a `<T extends SdkRequest>` and return `T`?

Reply:
I'll see what I can do. But, it's not that simple. The actual type is quite complex: `T extends S3Request.Builder, SdkPojo, CopyableBuilder<Builder, T>>` and the methods I need to use don't necessarily have the right return type either.

Reply:
I tried - not worth the effort.